CMakeLists.txt

```diff
@@ -115,8 +110,8 @@ else()
   set(DMLC_CMAKE_LITTLE_ENDIAN 1)
 endif()
 
-message(STATUS "${CMAKE_LOCAL}/build_config.h.in -> ${INCLUDE_DMLC_DIR}/build_config.h")
-configure_file("${CMAKE_LOCAL}/build_config.h.in" "${INCLUDE_DMLC_DIR}/build_config.h")
+message(STATUS "${CMAKE_LOCAL}/build_config.h.in -> include/dmlc/build_config.h")
+configure_file("cmake/build_config.h.in" "include/dmlc/build_config.h")
```

Comment:
Please use more specific path like `${CMAKE_CURRENT_LIST_DIR}/cmake/` or `${dmlc_SOURCE_DIR}/cmake/`

Reply:
Why do you think this is necessary? Documentation is very clear about there configure_file expects it’s inputs and outputs when using relative paths. I think this is nothing we need to expect to change in cmake. So basically trying to pin cmake behaviour using a lot of custom variables is somthing that should be avoided in favor of maintainable and readable code.

Reply:
@codingforfun I just checked the documentation for `configure_file()` and it says that all relative paths are interpreted with respect to the value of [`CMAKE_CURRENT_SOURCE_DIR`](https://cmake.org/cmake/help/v3.2/variable/CMAKE_CURRENT_SOURCE_DIR.html#variable:CMAKE_CURRENT_SOURCE_DIR). I'm not sure `CMAKE_CURRENT_SOURCE_DIR` is necessarily equivalent to [`PROJECT_SOURCE_DIR`](https://cmake.org/cmake/help/v3.2/variable/PROJECT_SOURCE_DIR.html).

Here are definitions:

* `CMAKE_CURRENT_SOURCE_DIR`: The path to the source directory currently being processed. This is the full path to the source directory that is currently being processed by cmake.
* `PROJECT_SOURCE_DIR`: Top level source directory for the current project. This is the source directory of the most recent `project()` command.

Currently, the `CMAKE_LOCAL` variable is defined in terms of `PROJECT_SOURCE_DIR`. I think that `PROJECT_SOURCE_DIR` is correct choice here, since we want the source directory of the most recent `project()` command (dmlc-core is included by xgboost via `add_subdirectory()` command.)

Reply:
`CMAKE_CURRENT_SOURCE_DIR` is always the directory the cmake file lives in from where you call it. So in this case of a CMakeLists living in the top level directory of dmlc, this is the same as `PROJECT_SOURCE_DIR`. Documentation is somewhat unclear here.

Reply:
The stackoverflow discussion was regarding `CMAKE_SOURCE_DIR` which is a differnent one.

Reply:
@codingforfun Let me try your change on top of XGBoost and see if it compiles.

Reply:
Indeed, it seems that `CMAKE_CURRENT_SOURCE_DIR` is practically same as `PROJECT_SOURCE_DIR` for our use case.